src/main/scala/is/hail/asm4s/CM.scala

```diff
@@ -1,20 +1,22 @@
-package is.hail.expr
-
-import is.hail.asm4s.{Code, _}
+package is.hail.asm4s
 
 import scala.collection.mutable
 import scala.reflect.ClassTag
 
 object preCM {
-  case class E(m: Map[String, (Type, Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
-  case class S(fa: Array[AnyRef], ec: EvalContext)
+  case class E(m: Map[String, (ClassTag[AnyRef], Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
+  case class S(fa: Array[AnyRef], st: AnyRef)
 
-  def emptyE = E(Map[String, (Type, Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
-  def emptyS(ec: EvalContext) = S(Array[AnyRef](), ec)
+  def emptyE = E(Map[String, (ClassTag[AnyRef], Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
+  def emptyS(ec: AnyRef) = S(Array[AnyRef](), ec)
```

Comment:
Correct: if we use Scala's binding structure, we will maintain type information (because we don't traffic our types into and out of a `Map` mediating their identity through `String`s).

I planned to move the environment off of `CM` and into the AST `compile` methods as a separate PR, but I can do it as a part of this one too, if you prefer that.